_posts/2021-01-20-the-go-interface.md

```diff
@@ -0,0 +1,376 @@
+as a by-product also declaratively limits what the app controls.
+
+```golang
+type OpenWriter interface {
+    Open() (closer, error)
+    Write(p []byte) (n int, err error)
+}
+```
+
+Each storage type needs to implement the methods that our interface defined.
+That implies that unlike the previous implementation, each storage should
+have a struct with methods bind to it.
+
+On the app side, we just need to replace the storeType type from a string to
+our newly interface and then just call the methods to prepare the storage and
+write on it.
+
+```golang
+func Save(data string, store OpenWriter) error {
+   close, err := store.Open()
+   if err != nil {return err}
+   defer close()
+
+   if _, err := store.Write(encodeData(data)); err != nil {
+       return err
+   }
+}
+```
+
+### What about tests?
+
+Testable code *requires* decoupling and isolation. The test is a user of the
+tested code and needs to control its dependent data.
+
+When the tested code depends on interfaces, the test can substitute its own
+concrete objects and control the flow.
+
+Our current `app.Save` function is already testable.
+Lets examine the flows we need to check:
+- Open() fails.
+- Write() fails.
+- Data saved successfully.
+
+To do so, we just need to build a concrete that implements the interface
+and at the same time allows the test to control the responses.
+
+Lets implement our test stub:
+
+```golang
+type testStore struct {
+    openFail  bool
+    writeFail bool
+}
+
+func (s testStore) Open() (closer, error) {
+    var close closer = func(){}
+    if s.openFail {
+        return close, fmt.Errorf("failed")
+    }
+    return close, nil
+}
+
+func (s testStore) Write(data []byte) (int, error) {
+    if s.writeFail {
+        return 0, fmt.Errorf("failed")
+    }
+    return nil
+}
+```
+
+And here are our tests (using ginkgo/gomega):
+
+```golang
+var _ = Describe("app Save", func() {
+    It("fail to open store", func() {
+        store := testStore{openFail: true}
+        Expect(app.Save([]byte{}, store)).NotTo(Succeed())
+    })
+
+    It("fail write to store", func() {
+        store := testStore{writeFail: true}
+        Expect(app.Save([]byte{}, store)).NotTo(Succeed())
+    })
+
+    It("succeed write to store", func() {
+        store := testStore{}
+        Expect(app.Save([]byte{}, store)).To(Succeed())
+    })
+})
+```
+
+## Please avoid these
+
+### Return an interface
+Please return a concrete type from a factory or a constructor, not an interface.
+
+Returning an interface hints that the concrete type exposes only a specific
+behavior which is used by its consumers.
+But this limits the concrete type to expand and support other behaviors (it is
+limited by the interface).
+
+It may also be a premature abstraction.
+
+But there are always exceptions. A factory that may return different concrete
+types must have an interface as the return value.
+
+### Define interfaces side by side with concrete objects
```

Comment:
ditto

Reply:
I can ref an example (`io` package). Expressing this in code using different packages will add a lot of code.

Reply:
I agree. If you can refer to an already existing example then it would be best.